litgpt/__main__.py

```diff
@@ -0,0 +1,137 @@
+                "help": "Generation script that partitions layers across devices to be run sequentially.",
+            },
+            "tp": {
+                "fn": generate_tp_fn,
+                "help": "Generation script that uses tensor parallelism to run across devices.",
+            },
+        },
+        "chat": {"help": "Chat with a model.", "fn": chat_fn},
+        "convert": {
+            "help": "Utilities to convert from and to LitGPT.",
+            "to_litgpt": {"fn": convert_hf_checkpoint_fn, "help": "Convert Hugging Face weights to LitGPT weights."},
+            "from_litgpt": {"fn": convert_lit_checkpoint_fn, "help": "Convert LitGPT weights to Hugging Face weights."},
+            "pretrained_checkpoint": {
+                "fn": convert_pretrained_checkpoint_fn,
+                "help": "Convert a checkpoint after pretraining.",
+            },
+        },
+        "download": {"help": "Download weights or tokenizer data from the Hugging Face Hub.", "fn": download_fn},
+        # TODO: should this be at the first level?
+        "merge_lora": {"help": "Merges the LoRA weights with the base model.", "fn": merge_lora_fn},
+    }
+
+    from jsonargparse import set_docstring_parse_options, set_config_read_mode
+
+    set_docstring_parse_options(attribute_docstrings=True)
+    set_config_read_mode(urls_enabled=True)
+
+    root_parser = _new_parser(prog="litgpt")
+
+    # register level 1 subcommands and level 2 subsubcommands. If there are more levels in the future we would want to
+    # refactor this to do BFS traversal for registration
+    subcommands = root_parser.add_subcommands()
+    subcommand_to_parser = {}
+    for k, v in parser_data.items():
+        subcommand_parser = _new_parser()
+        if "fn" in v:
+            subcommand_parser.add_function_arguments(v["fn"])
+        else:
+            subcommand_to_parser[k] = subcommand_parser
+        subcommands.add_subcommand(k, subcommand_parser, help=v["help"])
+    for subcommand, parser in subcommand_to_parser.items():
+        subcommands = parser.add_subcommands()
+        for k, v in parser_data[subcommand].items():
+            if k == "help":
+                continue
+            subsubcommand_parser = _new_parser()
+            subsubcommand_parser.add_function_arguments(v["fn"])
+            subcommands.add_subcommand(k, subsubcommand_parser, help=v["help"])
+
+    args = root_parser.parse_args()
+    args = root_parser.instantiate_classes(args)
+
+    subcommand = args.get("subcommand")
+    subargs = args.get(subcommand)
+    subsubcommand = subargs.get("subcommand")
+    subsubargs = subargs.get(subsubcommand) if isinstance(subsubcommand, str) else None
+
+    level_1 = parser_data[subcommand]
+    if subsubcommand is None:
+        fn = level_1["fn"]
+        kwargs = subargs
+    else:
+        fn = level_1[subsubcommand]["fn"]
+        kwargs = subsubargs
+    kwargs.pop("config")
+
+    torch.set_float32_matmul_precision("high")
```

Comment:
Did you put this here so it can be removed from the other scripts?

Reply:
No. I put this here because this CLI doesn't call the `__main__` of the scripts themselves. And that's where we are calling this. Maybe it should be moved into the script functions instead of in the `__name__ == "__main__"` block so that this CLI doesn't need to repeat it